pennylane/qchem/convert.py

```diff
@@ -354,3 +354,81 @@ def import_operator(qubit_observable, format="openfermion", wires=None, tol=1e01
         return qml.dot(*_openfermion_to_pennylane(qubit_observable, wires=wires))
 
     return qml.Hamiltonian(*_openfermion_to_pennylane(qubit_observable, wires=wires))
+
+
+def _excitations(electrons, orbitals):
+    r"""Generate all possible single and double excitations from a Hartree-Fock reference state.
+
+    Args:
+        electrons (int): number of electrons
+        orbitals (int): number of spin orbitals
+
+    Returns:
+        tuple(list, list): lists with the indices of the spin orbitals involved in the excitations
+
+    **Example**
+
+    >>> electrons = 2
+    >>> orbitals = 4
+    >>> _excitations(electrons, orbitals)
+    ([[[0, 2]], [[0, 3]], [[1, 2]], [[1, 3]]], [[0, 1, 2, 3]])
+    """
+    singles_p, singles_q = [], []
+    doubles_pq, doubles_rs = [], []
+
+    for i in range(electrons):
+        singles_p += [i]
+        doubles_pq += [[k, i] for k in range(i)]
+    for j in range(electrons, orbitals):
+        singles_q += [j]
+        doubles_rs += [[k, j] for k in range(electrons, j)]
+
+    singles = [[p] + [q] for p in singles_p for q in singles_q]
+    doubles = [pq + rs for pq in doubles_pq for rs in doubles_rs]
+
+    return singles, doubles
+
+
+def _excitated_states(electrons, orbitals, excitation):
+    r"""Generate excited states from a Hartree-Fock reference state.
+
+    Args:
+        electrons (int): number of electrons
+        orbitals (int): number of spin orbitals
+        excitation (int): number of excited electrons
+
+    Returns:
+        tuple(array, array): arrays of excited states and signs obtained by reordering of the creation operators
+
+    **Example**
+
+    >>> electrons = 2
+    >>> orbitals = 5
+    >>> excitation = 2
+    >>> _excitated_states(electrons, orbitals, excitation)
+    (array([28, 26, 25]), array([ 1, -1,  1]))
+    """
+    hf_state = qml.qchem.hf_state(electrons, orbitals)
+    singles, doubles = _excitations(electrons, orbitals)
+    states, signs = [], []
+
+    if excitation == 1:
+        for s in singles:
+            state = hf_state.copy()
+            state[s] = state[s[::-1]]
+            states += [state]
+            signs.append((-1) ** len(range(s[0], electrons - 1)))
+
+    if excitation == 2:
+        for d in doubles:
+            state = hf_state.copy()
+            state[d] = state[[d[2], d[3], d[0], d[1]]]
+            states += [state]
+            order_pq = len(range(d[0], electrons - 1))
+            order_rs = len(range(d[1], electrons - 1))
+            signs.append((-1) ** (order_pq + order_rs + 1))
+
+    states_str = ["".join([str(i) for i in state]) for state in states]
+    states_int = [int(state[::-1], 2) for state in states_str]
```

Comment:
combining into one list comprehension wouldn't hurt readability that much, but would eliminate an extra loop. Alternatively might be able to use `zip` to cast the strings to integers (I think it's faster than list comprehension?..)

Reply:
For the first option, we could just cast to int right away, just combining the int() call into the first list -- something like
```
states_str = [int("".join([str(i) for i in state])[::-1], 2) for state in states]
```
As for the second option, I played around with it for a bit and I don't think it'd actually be much faster -- plus it is a bit tricky to cast to int from base 2 string.

Reply:
The reason for having two separate lines is to make the function flexible for returning either the string representation or the int number representation. I think we need this flexibility in the future.

Reply:
Nice! Agreed, this will make it more scalable!